docs/source-2.0/guides/codegen/mapping-shapes-to-languages.rst

```diff
@@ -0,0 +1,629 @@
+
+.. code-block:: java
+
+   NullableIndex index = NullableIndex.of(model);
+   if (index.isNullable(someMember)) {
+       // optional
+   } else {
+       // always present
+   }
+
+
+FAQ
+===
+
+Should constraint traits impact generated types?
+------------------------------------------------
+
+In general, no. Baking anything about these traits into generated types
+makes these traits impossible to change in the future without breaking
+previously generated code.
+
+- Length and pattern traits should have no impact on generated types.
+  Shapes with a length or pattern trait should be represented as a
+  standard string type.
+- Range traits must have no impact on generated types. Code generators
+  must not rely on range traits to determine which numeric type is best
+  for representing a Smithy shape. Instead, code generators must rely
+  on the Smithy type used for the numeric shape (integer, long, etc).
+- The ``@enum`` trait was replaced by the ``enum`` shape in Smithy IDL
+  2.0. Both the ``@enum`` trait and ``enum`` shape can influence code
+  generation, though they must be considered a specialization of
+  strings. This allows servers to add new enum values over time without
+  breaking previously generated clients.
+- The ``@required`` trait is no longer considered a constraint trait in
+  Smithy IDL 2.0. It is now expected to influence code generated types.
+
+
+Should clients enforce constraint traits?
+-----------------------------------------
+
+No. A client should defer the validation of constraint traits to the
+service.
+
+
+Why don't we validate constraint traits on the client?
+------------------------------------------------------
+
+Validating constraint traits on the client makes it extremely hard to
+change constraint traits, even in what appears to be a backward
+compatible change. Changes to constraint traits need to be backward
+compatible by making the constraint more relaxed. However, things fall
+apart when multiple actors are creating resources using different
+versions of the service.
+
+Years ago, many AWS SDKs validated constraint traits client-side and
+refused to send non-compliant input. During that period, Amazon EC2
+updated instance IDs to support a longer string, and existing SDKs began
+to break when this change was deployed. That's because the use of other
+tools like the AWS Management Console or even managed services created
+instances that SDKs could no longer interact with. If a client
+encountered an instance created in the console and then tried to make a
+subsequent call using the instance ID, the client would refuse to send
+the request. This resulted in a large amount of customer pain and took
+months of effort to correct. If the client ignored constraint traits and
+allowed the service to enforce them, the change would have been
+transparent to previously generated clients.
+
+
+Do generators need to worry about mixins?
```

Comment:
This can be removed in lieu of "Mixins are an implementation detail of the model".

Reply:
Kept the FAQ entry to reiterate, but now just say "No." and link to the relevant section.